src/helpers.ts

```diff
@@ -0,0 +1,119 @@
+import type {
+  ColumnarVectors,
+  MakeRequestTiming,
+  RequestTiming,
+  Vector,
+} from "./types";
+
+export const runtime = globalThis.navigator?.userAgent.split("/")[0];
+export const isRuntimeFullyNodeCompatible =
+  runtime === "Node.js" || runtime === "Deno";
+
+/** An error class for errors returned by the turbopuffer API. */
+export class TurbopufferError extends Error {
+  status?: number;
+  constructor(
+    public error: string,
+    { status, cause }: { status?: number; cause?: Error },
+  ) {
+    super(error, { cause: cause });
+    this.status = status;
+  }
+}
+
+/** A helper function to determine if a status code should be retried. */
+export function statusCodeShouldRetry(statusCode?: number): boolean {
+  return !statusCode || statusCode === 408 || statusCode === 429 || statusCode >= 500;
+}
+
+/** A helper function to delay for a given number of milliseconds. */
+export function delay(ms: number) {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+export function make_request_timing({
+  request_start,
+  response_start,
+  body_read_end = 0,
+  decompress_end = 0,
+  deserialize_end = 0,
+  requestCompressionDuration = 0,
+}: MakeRequestTiming): RequestTiming {
+  return {
+    response_time: response_start - request_start,
+    body_read_time: body_read_end ? body_read_end - response_start : 0,
+    compress_time: requestCompressionDuration ? requestCompressionDuration : 0,
+    decompress_time:
+      decompress_end && body_read_end ? decompress_end - body_read_end : 0,
+    deserialize_time:
+      deserialize_end && (isRuntimeFullyNodeCompatible ? decompress_end : body_read_end)
```

Comment:
if we make decompress_end optional, this could use its presence to decide what to do rather than checking isRuntimeFullyNodeCompatible